ffcx/codegeneration/integrals.py

```diff
@@ -691,3 +688,29 @@ def generate_block_parts(self, quadrature_rule: QuadratureRule, blockmap: Tuple,
         quadparts += [pre_loop, hoist, body]
 
         return preparts, quadparts
+
+    def fuse_loops(self, definitions):
+        """
+        Merge a sequence of loops with the same iteration space into a single loop.
+        Loop fusion improves data locality, cache reuse and decreases the loop control overhead.
+        NOTE: Loop fusion might increase the pressure on register allocation.
+        Ideally, we should define a cost function to determine how many loops should fuse at a time.
```

Comment:
Yep, but I am not a big fan of doing these compiler-alike optimisations in FFCx codegen layer. Would be much cleaner to have a separate "compiler pass" for these.

Reply:
I agree that it's not ideal. But we would lose "locality". 
Parsing the entire code is much more complicated than applying optimizations to, say, coefficient computations.